packages/flutter/lib/src/painting/image_cache.dart

```diff
@@ -71,21 +120,46 @@ class ImageCache {
   ImageStreamCompleter putIfAbsent(Object key, ImageStreamCompleter loader()) {
     assert(key != null);
     assert(loader != null);
-    ImageStreamCompleter result = _cache[key];
-    if (result != null) {
+    final ImageStreamCompleter result = _pendingImages[key];
+    // Nothing needs to be done because the image hasn't loaded yet.
+    if (result != null)
+      return result;
+    final _CachedImage image = _cache.remove(key);
+    if (image != null) {
       // Remove the provider from the list so that we can put it back in below
       // and thus move it to the end of the list.
-      _cache.remove(key);
+      _cache[key] = image;
+      return image.completer;
     } else {
-      if (_cache.length == maximumSize && maximumSize > 0)
-        _cache.remove(_cache.keys.first);
-      result = loader();
+      final ImageStreamCompleter result = loader();
+      void listener(ImageInfo info, bool syncCall) {
+        final _CachedImage image = new _CachedImage(result, info.image == null ? 0 : info.image.height * info.image.width * 4);
+        _pendingImages.remove(key);
+        _cache[key] = image;
+        _checkCacheSize();
+        result.removeListener(listener);
```

Comment:
add a test to verify that this removeListener call works as you expect

Reply:
should be covered by the fixed test in packages/flutter/test/widgets/image_test.dart